Use childProcess.fork, add HMR support

wmertens opened the pull request:
Big overhaul, resulting in speedier hot reloads, fixes #12 and fixes #9.

Needs a major version bump: the configuration changed and HMR handlers need to be removed vs v1.

wmertens commented:
@tikotzky thoughts?

tikotzky commented:
Looks good 👍
Is it worthwhile giving a way to opt out of auto injecting the HMR monitor script?

2 other questions
1. Will it auto restart the process if its unable to apply a HMR patch?
1. Will it auto restart the process if the process dies for any other reason?

tikotzky commented:
Just confirming that while would replace #17, we'd still need something like #18 in order to allow manual restarts.

tikotzky commented:
Looking at this again, I see 2 changes happening in the PR.
1. a change to using `child_process.fork` instead of cluster.
1. doing HMR with a signal instead of needing to poll.

While I agree that this plugin should be switching to `child_process.fork`, I'm not sure if its best to add a custom loader and monitor script. The only benefit that I see over using `webpack/hot/signal` is that using this approach if HMR is unable to apply the patch you'll be able to restart the entire server. If #18 gets merged you'd be able to manually restart the server in the case that HRM is unable to be applied.

Is that single benefit worth the complexity being added to support it?

wmertens commented:
The monitor script also serves to know if the script loaded
successfully/applied HMR successfully. Without it, it will go in a restart
loop when it goes wrong.

As you note, it also quits the process on broken HMR, unlike the webpack
versions. That non-quitting often caused confusion here in our team because
it wasn't noticed and things start to misbehave. (we HMR client+server + do
SSR in dev)

Also, I had trouble getting webpack/hot to work because I was using
node-externals, and this works for everyone.

So I still think it's worth the extra complexity and possible maintenance
on webpack changes.

And yes, the keyboard shortcut for restarting is still a good addition.

Would you be able to test the branch and implement #18 on top of it?

On Mon, Feb 19, 2018, 12:58 AM Mordy Tikotzky, <notifications@github.com>
wrote:

> Looking at this again, I see 2 changes happening in the PR.
>
>    1. a change to using child_process.fork instead of cluster.
>    2. doing HMR with a signal instead of needing to poll.
>
> While I agree that this plugin should be switching to child_process.fork,
> I'm not sure if its best to add a custom loader and monitor script. The
> only benefit that I see over using webpack/hot/signal is that using this
> approach if HMR is unable to apply the patch you'll be able to restart the
> entire server. If #18
> <https://github.com/ericclemmons/start-server-webpack-plugin/pull/18>
> gets merged you'd be able to manually restart the server in the case that
> HRM is unable to be applied.
>
> Is that single benefit worth the complexity being added to support it?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ericclemmons/start-server-webpack-plugin/pull/19#issuecomment-366560116>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADWlqXDwAWdVHqPbZMZrg7o6RfEKDgPks5tWLkJgaJpZM4SJ4bI>
> .
>

wmertens committed:
Use childProcess.fork()

* use process messaging to communicate
* check for loaded before restarting
* needs `start-server-webpack-plugin/dist/monitor` to be loaded last in the entry
* make sure not to mark monitor as external, needs to be built with webpack

wmertens committed:
Auto-add monitor, refactor

wmertens committed:
cleaner logs

wmertens committed:
document changes

wmertens committed:
Fix tests

Feels a bit silly have to install webpack just to pass the tests, but maybe we can do better tests later.

Right now we don't test the instrumenting nor the running. Maybe we can test that end-to-end.

wmertens committed:
npm

wmertens committed:
options.keyboard => restartable

wmertens committed:
Add entry tests and test project

wmertens committed:
Simplify adding monitor

Instead of trying to find some hook, just insert the HMR module at plugin apply time

wmertens commented:
I think this is ready to become v3. There is even a test that checks if webpack runs the project :)

@tikotzky @RonanQuigley @ericclemmons thoughts?

wmertens committed:
v3.0.0-rc1

wmertens commented:
Published as `start-server-webpack-plugin@beta`, please test :)

wmertens commented:
I just realized that e.g. socket-cluster handles SIGUSR2 for restarting its workers, so the signal method should not be removed. I'll make it configurable.

wmertens committed:
Fix nodeArgs passing

if options.args wasn't defined, it caused .fork to ignore the third argument (options) and thus ignore `execArgv`.

wmertens committed:
configure prettier

wmertens committed:
publish v3rc2

wmertens committed:
Fix worker restart

wmertens committed:
lint/cleanup

wmertens committed:
Add signal option back

Useful for e.g. socket-cluster

wmertens committed:
monitor: log applied modules

wmertens committed:
publish v3-rc3

ericclemmons commented:
@wmertens Finally settled into Seattle!  What's the best way for me to try this out? 
> There is even a test that checks if webpack runs the project :)

Running the suite alone, or should I update one of my personal projects to try it all out?

wmertens commented:
@ericclemmons great!

You can simply install the @beta npm version, and if you need to you can
add the `entryName` of the server entry. That should be all.

The test is run by `npm test`

On Sat, Mar 24, 2018 at 7:51 PM Eric Clemmons <notifications@github.com>
wrote:

> @wmertens <https://github.com/wmertens> Finally settled into Seattle!
> What's the best way for me to try this out?
>
> There is even a test that checks if webpack runs the project :)
>
> Running the suite alone, or should I update one of my personal projects to
> try it all out?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ericclemmons/start-server-webpack-plugin/pull/19#issuecomment-375916127>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADWloeau1fahwjy9szXuqfAsFU3Ev00ks5thpXEgaJpZM4SJ4bI>
> .
>

ericclemmons commented:
@wmertens Was playing with it this morning. 🛳️

antonybudianto commented:
Hi,

I got when trying the beta

```
sswp> running `node /home/xx/react-ssr-starter/dist/bundle.css --`
sswp> Script did not load or failed HMR, not restarting
```

config:
```
entry: {
    bundle: project.paths.client('renderer/server.js')
},
```

```
new StartServerPlugin({
  entryName: 'bundle'
})
```

wmertens commented:
@antonybudianto for some reason, the bundle entry point is the css file. You can edit `node_modules/start-server-webpack-plugin/dist/StartServerPlugin.js` and add a `console.log(compilation.entrypoints)` around this line https://github.com/ericclemmons/start-server-webpack-plugin/blob/use-the-fork/src/StartServerPlugin.js#L60 (search for `entrypoints`) to see what's going on.

You can also misspell your entryName and then it will tell you the entries it sees, maybe that's a quicker check.

If we can't figure it out, we'll have to add a `scriptFile` option so you specify exactly what the script is that should be run; however it could be that the HMR addition won't have been added.

antonybudianto commented:
@wmertens I've tweaked my config and now it works (HMR on client + server)! https://github.com/antonybudianto/react-ssr-starter/pull/1

Looking forward the final release, thanks guys 👍

wmertens commented:
@antonybudianto great! is this something that was obvious or should we document this as a pitfall?

antonybudianto commented:
I just moved `mini-css-extract-plugin` from base webpack config into client webpack config, and it worked. Actually it used [name].css as the `fileName`, which becomes bundle.css.
Probably `scriptFile` option will be helpful or making sure the entry file is script file.

tikotzky commented:
Perhaps we should add a check and if the file is not a js file then print a nice error with instructions instead of letting it just fail.

antonybudianto commented:
I've used this in [cra-universal](https://github.com/antonybudianto/cra-universal) too, so far so good 👍

wmertens committed:
update scriptfile after every compilation

This allows using hash in the output filename

wmertens committed:
mocha: use babel-register

wmertens committed:
Merge branch 'master' into use-the-fork

ericclemmons commented:
Closing since this is 7 years old :D
